csrc/device_lower/pass/index.cpp

```diff
@@ -1816,6 +1821,224 @@ Val* hardCodedIndexGenerationForStMatrix(
   return out;
 }
 
+// Goal: Store (tma_m, tma_n) row-major tile in shared memory using stmatrix
+// (stsm_m_tile, stsm_n_tile)
+//
+// Let shared_memory tile be (tma_m = 64, tma_n = 128).
+// Let StMatrix tile be (stsm_m_tile = 16, stsm_n_tile = 16).
+// Let dtype be fp16 or bf16, so dtype_size is 2B.
+//
+// To avoid shared memory bank conflicts, apply swizzle on stmatrix and
+// tma store operations. Let swizzle be MmaInputSmemSwizzle::B128.
+//
+// For the TMA store, create box for 128B swizzle.
+// Let inner_tile_size = getBytesFromSwizzle(swizzle) / dtype_size = 64.
+//
+// Given shared memory tile [M(64), N(128)], split inner dimension by
+// inner_tile_size and reorder dimensions to get [NO(2), M(64), NI(64)].
+// The TMA Box is [M, NI]. Apply swizzle to the box using swizzleTMABox.
+//
+// Each ThreadIdx.y handles a (tma_m, tma_n) tile.
+// To account for the threadIdx.y, we have to add it to the offset:
+//   offset_from_tdy = threadIdx.y * tma_m * tma_n * 2 (half)
+//
+// Now, lets apply stmatrix tile to the TMA Box.
+// [NO(2), MO(4), MI(16), NIO(4), NII(16)].
+//
+// StMatrix is a warp-level operation. Four StMatrix operations can be issued
+// simultaneously given that a warp group is 128 threads. They are applied on
+// the MO dimension.
+//
+// [NO(2), MO(4) - TDX, MI(16) - StMatrix, NIO(4), NII(16) - StMatrix].
+// Virtually, there are 2 for-loops of size NO(2) and NIO(4).
+//
+// The Input TensorView for StMatrix after scheduleMmaOutputAllocation is
+// [128(TIDx), 8(n), 2, 2]. Given this domain, there is a serial for-loop of
+// size 8.
+//
+// Get indices for virtual for-loops given the serial for-loop:
+//   outer_index = loop->index() / NIO(4)
+//   inner_index = loop->index() % NIO(4)
+//
+// Each outer_index handles a (tma_m, NI(64)) tile.
+// To account for the outer_index, we have to add it to the offset:
+//   offset_from_outer_index = outer_index * tma_m * NI(64) * 2 (half)
```

Comment:
Can you add a line here explaining why it is
outer_index * tma_m * NI(64) * 2 (half)
 and not
 outer_index * NI(64) * 2 (half)

Reply:
The allocation domain of the shared memory tensor must match the loop domain of the global memory tensor. The loop domain is scheduled as `[NO(2), M(64), NI(64)]`. We must store the data in shared memory in `[M(64), NI(64)]` contiguous tiles.